views: Refactor popups to expedite the popup creation process.

preetmishra opened the pull request:
The intention is to extract and generalize duplicate code to expedite the popup creation process.

Along with that, it tidies up popups, that derive from PopUpView, by calculating their required width and height based on the content instead of using hard-coded values.

Tests amended.

As a side effect, fixes #654 and #658.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1308,8 +1308,7 @@ def test_keypress_exit_popup(self, key):
         assert self.controller.exit_popup.called
 
     def test_height_noreactions(self, message_fixture):
-        self.msg_info_view = MsgInfoView(self.controller, message_fixture)
```
This was redundant.

preetmishra commented:
I'll proceed to add the tests, for the new functions, if the implementation and the approach look good to everyone.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -775,6 +777,55 @@ def __init__(self, controller: Any, widgets: List[Any],
+        """
+        Return a list of widgets to be rendered and sets height accordingly.
+        """
+        col_max_widths = self.max_widths(contents, title_len, dividechars)
```
You're passing these values in (contents, title_len, dividechars), to use them in an interior method and only contents otherwise. How might this look if you didn't nest the calls, but called them sequentially?

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -807,37 +858,15 @@ def __init__(self, controller: Any,
+                (binding['help_text'], ', '.join(binding['keys']))
+                for binding in [binding for binding in KEY_BINDINGS.values()
+                                if binding['key_category'] == category]
+            ]))
```
You've compacted the generation of this data-structure, and while it's great that the data is more localized than it was when mixed with the UI elements, it's not quite as clear as it was before - it has a lot of nested elements.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -775,6 +777,55 @@ def __init__(self, controller: Any, widgets: List[Any],
+        self.width = max(heading_width, sum(col_max_widths) + dividechars)
+        return col_max_widths
+
+    def build_widgets(self, contents: PopUpViewContent, title_len: int,
```
I know that `title_len` is what is really passed in, but this value need not be the length of the title - what does it mean to pass in a certain value, for the purposes of these functions? Does that give you an idea for a better name?

preetmishra replied:
There are corner cases wherein the calculated width for the 'content' is smaller than the popup title. I use `title_len` to make sure that the calculated width is wide enough to display the popup title every time.

I will rename this to something better.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -878,50 +907,25 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        msg_info = [
+            ('', [('Date & Time', time.ctime(msg['timestamp'])[:-5])]),
+            ('', [('Sender', msg['sender_full_name'])]),
+            ('', [('Sender\'s Email ID', msg['sender_email'])]),
```
Is this intended as 3 separate categories with no name?

preetmishra replied:
Yes.

neiljp commented on zulipterminal/core.py:
```diff
@@ -109,16 +109,17 @@ def exit_popup(self) -> None:
     def show_stream_info(self, color: str, name: str, desc: str) -> None:
-        show_stream_view = StreamInfoView(self, color, name, desc)
+        show_stream_view = StreamInfoView(self, color, desc, len('# ' + name))
         self.show_pop_up(show_stream_view, "# {}".format(name))
```
Is there a reason you didn't tidy this title duplication in addition to the other title changes?

preetmishra replied:
Thanks for pointing it out. I missed it.

preetmishra commented:
@neiljp Thanks for the review. I have restructured the commit history to separate fixes and new changes from refactors, and also to make the refactors incremental.

A few comments about the commits:
* The first two commits are small fixes that could be squashed but I have left them separate for review.
* The next two commits extract `make_chart` (suggestions are welcome for the name) and `charts_width` from `HelpView` and `StreamInfoView`. These static methods are solely useful for popups that want to have or already have a similar chart-like structure.
* The subsequent commit introduces a new static method `get_popup_height` to dynamically calculate popup height that should be useful for any popup.
* The second last commit uses these new methods to simplify `MsgInfoView`.
* The last commit should be self-explanatory.

Re `__init__` vs `make_chart` (previously `build_widgets`), `make_chart` would only be useful for popups that want to have a similar chart-like structure which is why I have kept it separate.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -775,6 +779,34 @@ def __init__(self, controller: Any, widgets: List[Any],
+        """
+        Returns a list of widgets to render a chart.
+        """
+        strip_no = 0  # To track strips (do not increase for categories).
```
Should this be in the loop? I can't test now due to the czo reactions issue, but does the help menu look the same as before?
I think that was the intention of the `enumerate`.

preetmishra replied:
Earlier, in the help menu, the style started alternating from every 'category'. For instance,
```diff
Category 1
+
-
+

Category 2
+
-
+
```

However, now the style starts from the first strip and keeps alternating for every strip. For instance,
```diff
Category 1
+
-
+

Category 2
-
+
-
```

In addition to that, the new alternate styling is needed for the change. Otherwise, consecutive stripes that don't have categories would end up looking like:
```diff
+
+
+

Category 1
+
-
+
```
Instead of:
```diff
+
-
+

Category 1
-
+
-
```

`+`: `popup_constrast`
`-`: `None`

neiljp replied:
My preference is for the existing consistent line style after the category title; are you changing the behavior since you prefer the alternating style, or because the code is currently simpler with the un-categorized style, and results in that change as a side-effect?

Can we retain the original behavior if we group the '+++' lines into one category instead?

preetmishra replied:
I only changed the original behaviour to avoid having '+++' lines for stripes that do not have categories. 

Yes, grouping them into one category would work. Thanks!

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -814,28 +846,17 @@ def __init__(self, controller: Any) -> None:
+                                     ', '.join(binding['keys'])))
+            help_menu_content.append((HELP_CATEGORIES[category], key_bindings))
 
+        help_menu_content = PopUpView.make_chart(help_menu_content, max_widths)
```
Although this is static, you can still use `self.make_chart` to access the method. For the same reason as using `super()`, that avoids coupling in case the name changes.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -775,6 +779,34 @@ def __init__(self, controller: Any, widgets: List[Any],
+        for category, content in contents:
+            if category:
+                if strip_no != 0:
+                    widgets.append(urwid.Text(''))  # Add a blank line.
```
It's worth noting when/why; this was in the original code?

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -880,8 +880,11 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        # Add 4 (for 2 Unicode characters on either side) to the popup title
+        # length to make sure that the title gets displayed even when the
+        # content is shorter than the title length (+4 Unicode characters).
+        self.width = max(len(desc), len("# {}".format(name)) + 4)
```
Yes, it is how the bug gets fixed.

preetmishra replied:
Yes, that is how the bug gets fixed.

neiljp replied:
Well, this has two effects, as I understand it:
* it fixes the bug with an empty description, where the title isn't listed (?)
* it removes a border of one space either side where there are descriptions

I understand that this is a bugfix for the former, but the latter isn't obvious.

To clarify, we don't always have left/right padding in popups, but I'm not sure if we want to change this behavior for now?

preetmishra replied:
I understand the concern now. I will edit the commit to only do the former i.e. the bugfix. However, since the help menu doesn't add extra spaces, the two extra spaces will eventually get removed in the refactoring commit (`chart_widths`). Would that be okay?

preetmishra commented:
Yes, I left `MsgInfoView` because the refactors would end up 'changing' its behaviour and structure. However, `StreamInfoView` was straightforward to include.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -880,13 +880,13 @@ def keypress(self, size: urwid_Size, key: str) -> str:
         self.width = max(len(desc), len("# {}".format(name)) + 4)
-        self.height = 2
-        stream_info_content = [urwid.Text(desc, align='center')]
+        self.height = description.rows((self.width, ))
```
Thanks! I'll fix it.

preetmishra commented:
@neiljp Thanks for the feedback. I have updated the PR to address the suggestions that you left. 

The PR only involves `HelpView` in the refactoring commits and addresses other popups in separate commits as the extracted methods change their behaviours to some extent (which are mentioned in the commit message).

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -945,40 +944,26 @@ def __init__(self, controller: Any, color: str, desc: str,
+            ('', [('Date & Time', time.ctime(msg['timestamp'])[:-5]),
+                  ('Sender', msg['sender_full_name']),
+                  ('Sender\'s Email ID', msg['sender_email'])]),
+        ]
```
I have grouped these into a category as per the discussion in the other comment.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -779,6 +779,35 @@ def __init__(self, controller: Any, widgets: List[Any],
+            category_width = max(category_width, len(category))
+            for row in content:
+                if isinstance(row, str):
+                    text_width = max(text_width, len(row))
```
Same point here as the previous commit; HelpView doesn't use this?

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -838,13 +867,7 @@ def __init__(self, controller: Any,
-        max_widths = [max(width) for width in zip(*widths)]
-        self.width = sum(max_widths)
-
+    def __init__(self, controller: Any, popup_title: str) -> None:
```
Do we actually need to pass in the title? Didn't you settle on the length, previously? This is like a `minimum_width` parameter. That said, I see the appeal of just expecting the title, as then we can manage the extra spacing internally (eg. the +4)

preetmishra replied:
The reason I didn't go with `minimum_width` is that even when we pass `minimum_width=len(popup_title)`, we need the +4 in the `table_widths` method which doesn't justify the 'minimum width' name?

An alternative could be passing `len(popup_title) + 4` but that wouldn't be clear either (why we're adding +4 to the length?). To address that, we could add a comment but that'd require adding a comment before every call.

The approach where we handle the +4 internally allows adding a single comment.

I could use a suggestion about how to approach this.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -779,6 +779,35 @@ def __init__(self, controller: Any, widgets: List[Any],
+        column_widths = [max(width) for width in zip(*strip_widths)]
+
+        popup_width = max(title_width, category_width, text_width,
+                          sum(column_widths) + dividechars)
```
This is actually different behavior - arguably better and more general, but again not just a refactor!
(some of that comes from the other extra feature, ie. the `text_width`, of course)

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -814,28 +845,18 @@ def __init__(self, controller: Any) -> None:
+                                     ', '.join(binding['keys'])))
+            help_menu_content.append((HELP_CATEGORIES[category], key_bindings))
+
+        help_menu_content = self.make_table_with_categories(help_menu_content,
```
This reassigning is a bit confusing; it's perfectly valid python, and strictly `Sequence[...]` is compatible with `List[Any]` (can you see where I got those from?), but just using a different name would probably make this clearer :)

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -883,8 +891,7 @@ def __init__(self, controller: Any, popup_title: str) -> None:
                                                             column_widths)
-        self.height = len(help_menu_content)
-
+        self.height = self.get_popup_height(help_menu_content, popup_width)
```
As per the other comment re `help_menu_content` being re-used, using a different name would emphasize why the order is important here.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -945,40 +944,26 @@ def __init__(self, controller: Any, color: str, desc: str,
+                (reaction['emoji_name'], reaction['user']['full_name'])
+                for reaction in msg['reactions']
+            )
+            msg_info.append(('Reactions', reactions))
```
I like the reactions in a different category - that seems clearer for sure - but what do you think of grouping responses by alternating color line? Currently (in 0.5.1) it's all one color, which is the other extreme. I see we could do this by reaction, or *maybe* even by user? Grouping would also mean we could avoid having 'tada' listed however many times, for example!

If this fits within the current feature of the revised popup code, we could make that change in this commit, but otherwise we could delay this until later.

preetmishra replied:
@neiljp With the current (PR) code, one way to group could be using the reaction name for the first reactor and then leaving the subsequent left column blank for other reactors (with alternating `popup_contrast` style).

For instance,
```
+1       User 1 (popup_contrast)
         User 2
         User 3 (popup_contrast)
tada     User 1 
         User 2 (popup_contrast)
         User 3 

```
```python
[('+1', 'User 1'), ('', 'User 2'), ('', 'User 3'), ('tada', 'User 1'), ('', 'User 2'), ('', 'User 3')]
```


Does the proposed styling look good or should I make amendments to the `make_table_with_categories` code to incorporate the other behaviour (grouping responses by alternating colour line) that you suggested?

neiljp replied:
This is one option. I think originally I was considering you could have the users joined with newlines, so the above would be more like:
```
+1    User 1 (popup_contrast)
      User 2 (popup_contrast)
      User 3 (popup_contrast)
tada  User 1
      User 2
      User 3
smile User 4 (popup_contrast)
```
I thought this might use the contrast to group the reactions better, perhaps.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -932,15 +932,15 @@ def keypress(self, size: urwid_Size, key: str) -> str:
-        stream_info_content = [description]
+    def __init__(self, controller: Any, color: str, desc: str,
+                 popup_title: str) -> None:
+        stream_info_content = [('', [desc])]
```
Is this the commit where we'd introduce the one-string form of content? (ie. one column)

preetmishra replied:
Yes, this seems appropriate.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -932,15 +932,15 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        self.width = popup_width
+        stream_info_content = self.make_table_with_categories(
+            stream_info_content, column_widths)
+        self.height = self.get_popup_height(stream_info_content, popup_width)
```
Do we actually need to use this here? I suppose we will need it if we enhance this popup.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -932,15 +932,15 @@ def keypress(self, size: urwid_Size, key: str) -> str:
-        # Add 4 (for 2 Unicode characters on either side) to the popup title
-        # length to make sure that the title gets displayed even when the
-        # content is shorter than the title length (+4 Unicode characters).
-        self.width = max(len(desc), len("# {}".format(name)) + 4) + 2
```
Does this change match the refactoring? Another reason that moving the one-column code  into this commit might make sense is that the `len(desc)` code is replaced by the `max(textwidth, ...)` code? So what we use in the first commit should translate to the changes here.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -880,8 +880,11 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        # Add 4 (for 2 Unicode characters on either side) to the popup title
+        # length to make sure that the title gets displayed even when the
+        # content is shorter than the title length (+4 Unicode characters).
+        self.width = max(len(desc), len("# {}".format(name)) + 4) + 2
```
Yes, the +4 adds to the title length (to compensate for those 4 Unicode box drawing characters). I initially removed the +2 but that changed the original behaviour i.e. removed the two extra spaces.

Re 'desc+2, name+4', I think the +2, in the current iteration, is clearer, it indicates that we are explicitly adding 2 extra spaces to the minimum required width returned by the `max` function.

neiljp replied:
The issue is that we don't want to unconditionally add 2, as this could end up with len(title)+6.

preetmishra replied:
Ah, right. `max(len(desc) + 2, len("# {}".format(name)) + 4)` would be better. Thanks!

preetmishra commented:
@neiljp Thanks for the review and the suggestions, splitting and arranging the commits in an incremental order has definitely made it easier to comprehend the changes and spot any new changes.

I have moved code and updated the commits to introduce changes only when they're required and also removed 'extra' code from the refactor commits.

I have added a separate commit for `StreamInfoView` that essentially incorporates single column/text logic in the existing methods. We could either hold it off or merge it but I think it'd be better to include it while we're at it.

Re grouping reactions together, I second that. I would like to address it in a comparatively small follow-up PR.

preetmishra commented:
Updated with `max(len(desc) + 2, len("# {}".format(name)) + 4)`.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -778,6 +778,23 @@ def __init__(self, controller: Any, widgets: List[Any],
+    @staticmethod
+    def table_widths(contents: PopUpViewTableContent,
+                     dividechars: int=2,
+                     extra_width: int=0) -> Tuple[int, List[int]]:
```
It isn't used anywhere for now.

I added for making it easier to add extra width if we ever want to. Should we remove it and add it only when it is needed?

neiljp replied:
Let's leave it for now, unless we use it - actually, could it be used for padding the message info?

preetmishra replied:
It can only be used to increase the width from the inside - the content in the left column will be still left-aligned and the content in the right column will be still right-aligned. I will leave it for now.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -925,16 +933,16 @@ def keypress(self, size: urwid_Size, key: str) -> str:
-        # length to make sure that the title gets displayed even when the
-        # content is shorter than the title length (+4 Unicode characters).
-        self.width = max(len(desc) + 2, len("# {}".format(name)) + 4)
-        description = urwid.Text(desc, align='center')
```
While centered and padded content will look good as a standalone item, it will a bit odd in popups which also have two columns (where the left column is left and the right column is right aligned).

Re 'snug', I agree. I see that you've suggested 'mimicking the existing code for now' (which I assume means using the left column for text and leaving the right column blank). 

On another note, the `StreamInfoView` looks good without the changes that are being introduced in this commit. We could also hold off this commit as "we don't know what we'll need in the future".

Thoughts?

neiljp replied:
I was meaning to use padded and centered style on the single-column version, ie. the specialized version introduced in that commit, to retain the existing style. So two columns => help menu style, text => StreamInfo style (as it is now).

We could leave this commit for a later merge in this PR or elsewhere. Once things are tidied from the other comments I could merge some commits and we could either work on the remaining commit(s) here or move them to another branch/PR.

preetmishra replied:
I understand what you're suggesting. Though, I am concerned about the scenario where we would want to have both in the same popup i.e. the help menu style and the stream info style. It would look a bit off to have some content centre-aligned and some left/right-aligned. That said, if we decide to stick with either one of the styles in any popup (not mix single-column and two-column), that would work.

preetmishra commented:
@neiljp Thanks for the review. I have incorporated the in-line suggestions (leaving the ones that are related to `StreamInfoView` for now) and split the MsgInfoView commit into three parts.

Re grouping the reactions, I have left a proposal in https://github.com/zulip/zulip-terminal/pull/659#discussion_r432652509. I'll make the amendments as per your suggestion.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -880,13 +880,13 @@ def keypress(self, size: urwid_Size, key: str) -> str:
-        self.height = 2
-        stream_info_content = [urwid.Text(desc, align='center')]
+        description = urwid.Text(desc, align='center')
+        self.height = description.rows((self.width, ))
```
I was just going to merge this, but realized we have a subtle nuance here, as `self.width` is always requested to be big enough to hold the description, so we end up wit ha height of 1. The popup can be scrolled, but that's probably not what we want!

Given that the height is only known when showing the popup, we likely want to use a different approach for cases where we want some kind of minimum height visible, or use the column width we know we get in the popup to calculate it. The controller popup code needs improving, whether a special case for a 'minimum height visible' case like this, or a general change where the height is calculated later, maybe all from rows?

As part of this, we may wish to limit the maximum popup width, like we do with popup height?

preetmishra commented:
@neiljp I have updated the PR with the new height calculation approach (that we discussed in the PM) and with grouped reactions. 

Re `MsgInfoView`, since the original commit has been split into three, you would come across a few strange behaviours if you review it commit by commit. I have explained these [here](https://chat.zulip.org/#narrow/pm-with/2406,10457,14317-group/near/891242). We could squash these three commits post review.

preetmishra commented:
Updated to fix the commit order on GitHub.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -772,10 +772,11 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
 class PopUpView(urwid.ListBox):
     def __init__(self, controller: Any, widgets: List[Any],
-                 command: str) -> None:
+                 command: str, width: int) -> None:
```
I like having this first commit early, but maybe also have the last commit just afterwards, which applies the same logic to the title?
Then you can promptly follow it with the 'Consider category and title width...' commit, which means at least one of the latter commits doesn't break, in terms of not allowing space for the window title?

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -942,27 +941,14 @@ def __init__(self, controller: Any, msg: Message) -> None:
+                  ('Sender', msg['sender_full_name']),
+                  ('Sender\'s Email ID', msg['sender_email']),
+                  ('Reactions', reactions if msg['reactions']
+                   else '---None---')]),
```
While I know that this code is replaced when you use a category instead, I'm a little concerned that the layout code for columns doesn't work in general, in case we end up with a wide column again? Or is something else happening here?

preetmishra replied:
In this particular case, `reactions`, the data structure, that is being passed isn't what `make_table_with_categories` expects. 

However, since the old code uses a hardcoded width (`column_widths[1]`), this is likely to break for wide columns.

A fix could be to fall back `('weight', 1, urwid.Text(data))` whenever the maximum allowed width is smaller than `column_widths[1]`. Though, it would require `max_cols` to be passed into the method.

I couldn't come up with a way to pass maximum allowed width into `calculate_table_witdhs` without tampering the 'flow' of the commits in this PR.

I was thinking of passing a `maximum_width` argument, where `maximum_width=controller.maximum_popup_dimensions()[0]` and then making further amendments to `calculate_popup_width` and `make_table_with_categories` accordingly.

neiljp replied:
This seems to be mostly fixed by the updated commit in any case :+1:

preetmishra commented:
@neiljp Thank you for the review. I have made the proposed amendments and restructured the PR as per your in-line comments.

However, I couldn't address the column breakage bit in the `MsgInfoView`. I have left my findings in https://github.com/zulip/zulip-terminal/pull/659#discussion_r434257471. It would be really helpful if you could share a few pointers about the same.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -813,7 +813,10 @@ def calculate_table_widths(contents: PopUpViewTableContent,
-                strip_widths.append(list(map(len, row)))
+                # Replace text with the longest string if the text is seperated
+                # by newlines.
+                row_max = map(lambda text: max(text.split('\n'), key=len), row)
```
I don't have any preference for `map` or `lambda` syntax as such. It's better to use them in small doses, I agree.

I have combined the proposed implementation with the other `len` in the latest update. Thanks.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1,6 +1,6 @@
 import threading
 import time
-from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
```
This change amends the +2 extra width and the alignment that we don't have any test for. The commit doesn't change any other functionality.

preetmishra commented:
@neiljp Thanks for the review and the in-line suggestion.

I have updated that commit with `comprehension` syntax and improved the commit messages. 

Furthermore, I have moved the https://github.com/zulip/zulip-terminal/pull/659#discussion_r438341690 change into that other commit (where we had an issue with reactions) and used `''.join()` to make `reactions` compatible with `make_table_with_categories`. This should resolve the issue (partially).

Yes, I gather it would be better to hold the last commit off for now. I have removed it from the PR. I will rebase and push a WIP PR after we have merged these changes.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -814,7 +813,13 @@ def calculate_table_widths(contents: PopUpViewTableContent,
+                    len(max(text.split('\n'), key=len))
+                    for text in row
+                ]
+                strip_widths.append(max_row_lengths)
```
This looks clearer in this commit than the version with lambdas before! This does seem to fix the issue we saw earlier with this commit, unless the window/width gets super-narrow, which is probably going to be an issue to use ZT in any case.
